### PR TITLE
feat(ui): inline virtual text (2)

### PR DESCRIPTION
**Pull request by tom-anders**

### CODE REVIEW

**Comment by bfredl** [comment]
How does this relate/compare to @sleepySwords branch? https://github.com/SleepySwords/neovim/tree/inline2 

I appreciate the effort, now lets coordinate our work and make a plan for what we want to merge for an MVP.

**Comment by tom-anders** [comment]
> How does this relate/compare to @SleepySwords branch? https://github.com/SleepySwords/neovim/tree/inline2
> 
> I appreciate the effort, now lets coordinate our work and make a plan for what we want to merge for an MVP.

Sorry, I only posted this as a draft for now to see which issues with CI still need to be fixed.

**Comment by tom-anders** [comment]
@bfredl Still, here's a short overview of the commits currently present in this PR:
 
1. 9abd8a77d4a942577e4950d369489f7aaebfda65 feat(ui): inline virtual text
    - This is basically the content of #20130, plus a small fixup to make clang-tidy happy
2. 7ea02b7f5e50b061621ec6fe55b1e8691411756f Fix fold highlighting (broken by inline virtual text commit)
    - Fixes the issue described in #20524 (fold highlight being broken)
3. dbd4eae2c609b3859b171876061315d92c4bd689 fix(api): Disallow inline virtual text for ephemeral marks
    - Does what it says, as mentioned in your comment in the original PR
4. 6e68a902423fb8bc1ac79a3572f59d407b4c89a4 fix(ui): adjust the cursor when inserting virtual text
    - Taken from https://github.com/neovim/neovim/pull/20130#issuecomment-1369652743, without this fix editing becomes really annoying because the cursor jumps around a lot. 

So 1) is basically your original draft PR, 2) and 4) are taken from @SleepySwords' branch. 3) Is only in this branch.

IMO the best course of action would be to get this PR to a state where all tests pass, probably by cherry-picking some more commits from @SleepySwords' branch. Then we could merge this PR, mark this feature as experimental and go over the remaining vim patches, merging them  one-by-one if possible. What do you think @bfredl @SleepySwords ?

**Comment by SleepySwords** [comment]
That sounds good to me. I'm not sure about merging just as is though, I'd rather have this PR at least be usable enough to be used for something like Inlay Hints.

Also, tests should pass for the latest commit on my branch, all the relevant tests do pass for me while running `make functionaltest`, however, it does hang for me towards the very end.

File: src/nvim/decoration.h [hunk 1]
```diff
@@ -23,9 +23,12 @@ typedef enum {
   kVTOverlay,
   kVTWinCol,
   kVTRightAlign,
+  kVTInline,
 } VirtTextPos;
 
-EXTERN const char *const virt_text_pos_str[] INIT(= { "eol", "overlay", "win_col", "right_align" });
+EXTERN const char *const virt_text_pos_str[] INIT(=
```
Comment (bfredl) [inline src/nvim/decoration.h:29]: no line break between `=` and `{`

**Comment by bfredl** [comment]
@SleepySwords of the more patches you looked into and ported, which ones would you say are essential for inline hints? ("all of them" is a fine answer if that is the case :)

File: src/nvim/decoration.h (hunk 1 above)
Comment (tom-anders) [reply src/nvim/decoration.h:29]: Hmm, but this is the formatting applied by uncrustify, otherwise the linter will complain

**Comment by SleepySwords** [comment]
## Essential:
- 652d481c1a2e62c5d1c21c37290fb8871375e17f - fixes cursor position with multiple virtual text in the same position
- 74065c4a88450bfc7e618b4972de8ff8573cac40 - fixes cursor position when entering insert mode
- d6f21f8ebe4fc87e51aecb2230bafd344152429f - diff to `charset.c` is essential, the other bit fixes **multibyte** which may not be essential
- a186acefbcfb6ab71076a86c1435912c336790fc - fixes wrapping of an **overly long** virtual text, which may be considered **non-essential**
- 6ffa5ac09b09189a18be2a7959e53b3a030c994e - fixes offset when using hard tabs
- a569c2a7337c44ecca274a0f2f306c1d1a78e27e - not entirely sure how many people use linebreak, if it's a small enough number, this could be considered **non-essential**
- 45c5ac5c4f9eb04c78a83a235ce61a29e180108a - fixes virtual text being highlighted if character after is highlighted, via search or match
    - b5b26e786b6b6e12d4d95083179dd6894923ba99 - properly implements this
- c2e63214bda445c0824aca187678d6c54c4f2e42 - fixes cursor position jumping when inserting text if right_gravity is enabled
- ff513ffd812fa73d10053fedd97030930b5a278e - fixes weird rendering of virtual text with `nowrap` and virtual text is hidden
- 679d680a2160e5c0a679fc7794dd93a47fec50e1 - incorrect highlighting with virtual text with `set number`, (don't remember if this was for **overly long** lines)
- c2a447dd17a15dc634929787dd8bd60356348a19 - this is a prerequisite for the next commit, I can't remember if this commit causes the bug which the next commit fixes, or if it was always like this and needed to be implemented. I believe this PR was originally for highlighting leaking when it is **overly long**.
   - f503b0af58750910efb9254ddafa02731545716b - fix incorrect highlighting if next to a match

## Non essential
- 81f003df0b45cc533f0c7911063225fd37edd0e5 - wouldn't say this is essential, but it's a one-liner that puts the cursor at 0  when there is only virtual text, should be easy to implement
- 602c5b5aa36ebed9b95a5a87d8ab696a335c9db6 - was part of a larger irrelevant commit, IIRC this is only a refactor and should not affect functionality
- 4ecb54c16f5d8b0bceadf85cf90e14d032171379 - fixes an edge case where iirc fixes an **overly long** with `nowrap` enabled causes weird renderind
    - 7d94ab93af7a3e6405f9bc2c5a8f5a9c98f0e9e0 - fixes multibyte characters from implementing the above PR
    - cc66beacba68657bd1409cac581890dc1f918681 - The above PR, 4ecb54c16f5d8b0bceadf85cf90e14d032171379, sets `n_extra` to zero without selecting the next virtual text causing incorrect first character to be drawn
- d483e392339bd53e150b340eb41d1066fb6fbdb4 - only for virtual text in diff mode
- e844e92307acad859d541003dedb34c986427759 - only for virtual text in diff mode
- 6cb724d304e2328b061368f1640809d91696eff4 - very specific edge case regarding tabs and **overly long** virtual text
- cc4f9e0e234490eb38e5821483ffb61af99efafb - linting (obv not required)

## Already implemented
- 1e5f1a27307571b20a8ad0f705b89aba3ed8d9fe
- 98ac518f74f6d8352a266acd448672b57ae1a9b2
- d39f108af3e466aad52ca322bc3c66f4092044db
- bc7e00548a816c82401be49bed1115345454c02e

This covers all of the commits in my branch, if there is anything else I find, I'll probably post a comment here.

Some things that should probably be discussed:
- Should fixes to weird bugs introduced by overly long virtual texts be part of this PR?
- Should fixes to rendering multibyte characters be part of this PR?
- Should fixing up virtual text in diff mode be part of this PR?

@bfredl @tom-anders thoughts?

File: src/nvim/decoration.h (hunk 1 above)
Comment (lewis6991) [reply src/nvim/decoration.h:29]: Are you sure it was the right version of uncrustify with our config file?

**Comment by tom-anders** [comment]
> This covers all of the commits in my branch. There may be some vim-patches I have missed, but it works find on my end, if there is anything else I find, I'll probably post a comment here :)
> 
> If the patch came from a vim-patch (majority), it should be listed in the description which patch it came from.

Wow, thanks for the detailed overview! I'll try and cherry pick at least all the patches marked as "essential"

> Some things that should probably be discussed:
> 
>     * Should fixes to weird bugs introduced by overly long virtual texts be part of this PR?
> 
>     * Should fixes to rendering multibyte characters be part of this PR?
> 
>     * Should fixing up virtual text in diff mode be part of this PR?

I'll pass this decision on to @bfredl :)

File: src/nvim/decoration.h (hunk 1 above)
Comment (tom-anders) [reply src/nvim/decoration.h:29]:
```
$ uncrustify --version                                                       
Uncrustify-0.76.0
~~
while the config file says `Uncrustify-0.76.0_f` - is the `_f` suffix important?

But the CI also seemed to have the same issue: https://github.com/neovim/neovim/actions/runs/4855242666/jobs/8653579277#step:14:156

**Comment by bfredl** [comment]
> * Should fixes to rendering multibyte characters be part of this PR?

Is this whenever there are multibyte chars in a inline virtual text? I'd say we want this already.

**Comment by SleepySwords** [comment]
> > * Should fixes to rendering multibyte characters be part of this PR?
> 
> Is this whenever there are multibyte chars in a inline virtual text? I'd say we want this already.

yea it is, alright cool

**Comment by tom-anders** [comment]
So I went through @SleepySwords branch, cleaning up and squashing the commits. 

Right now, all the essential commits are included, as well as the non-essential ones - But I tried to move the
non-essential ones as far as possible to the back of the commit history, so that we can easily drop them if required.

@bfredl Anything we want to drop from the non-essential list?

## Essential

- [x] [652d481](https://github.com/neovim/neovim/commit/652d481c1a2e62c5d1c21c37290fb8871375e17f) - fixes cursor position with multiple virtual text in the same position

- [x] [74065c4](https://github.com/neovim/neovim/commit/74065c4a88450bfc7e618b4972de8ff8573cac40) - fixes cursor position when entering insert mode

- [x] [d6f21f8](https://github.com/neovim/neovim/commit/d6f21f8ebe4fc87e51aecb2230bafd344152429f) - diff to `charset.c` is essential, the other bit fixes **multibyte** which may not be essential

- [x] [a186ace](https://github.com/neovim/neovim/commit/a186acefbcfb6ab71076a86c1435912c336790fc) - fixes wrapping of an **overly long** virtual text, which may be considered **non-essential**, the change to the condition in `plines.c` may be essential, have not tried it out without that change.

- [x] [6ffa5ac](https://github.com/neovim/neovim/commit/6ffa5ac09b09189a18be2a7959e53b3a030c994e) - fixes offset when using hard tabs

- [x] [a569c2a](https://github.com/neovim/neovim/commit/a569c2a7337c44ecca274a0f2f306c1d1a78e27e) - not entirely sure how many people use linebreak, if it's a small enough number, this could be considered **non-essential**

- [x] [45c5ac5](https://github.com/neovim/neovim/commit/45c5ac5c4f9eb04c78a83a235ce61a29e180108a) - fixes virtual text being highlighted if character after is highlighted, via search or match
  
  - [x] (squashed into previous) [b5b26e7](https://github.com/neovim/neovim/commit/b5b26e786b6b6e12d4d95083179dd6894923ba99) - properly implements this

- [x] [c2e6321](https://github.com/neovim/neovim/commit/c2e63214bda445c0824aca187678d6c54c4f2e42) - fixes cursor position jumping when inserting text if right_gravity is enabled

- [x] [ff513ff](https://github.com/neovim/neovim/commit/ff513ffd812fa73d10053fedd97030930b5a278e) - fixes weird rendering of virtual text with `nowrap` and virtual text is hidden

- [x] [679d680](https://github.com/neovim/neovim/commit/679d680a2160e5c0a679fc7794dd93a47fec50e1) - incorrect highlighting with virtual text with `set number`, (don't remember if this was for **overly long** lines)

- [x] [c2a447d](https://github.com/neovim/neovim/commit/c2a447dd17a15dc634929787dd8bd60356348a19) - this is a prerequisite for the next commit, I can't remember if this commit causes the bug which the next commit fixes, or if it was always like this and needed to be implemented. I believe this PR was originally for highlighting leaking when it is **overly long**.
  
  - [x] [f503b0a](https://github.com/neovim/neovim/commit/f503b0af58750910efb9254ddafa02731545716b) - fix incorrect highlighting if next to a match


## Non essential

- [x] [81f003d](https://github.com/neovim/neovim/commit/81f003df0b45cc533f0c7911063225fd37edd0e5) - wouldn't say this is essential, but it's a one-liner that puts the cursor at 0  when there is only virtual text, should be easy to implement

- [x] [602c5b5](https://github.com/neovim/neovim/commit/602c5b5aa36ebed9b95a5a87d8ab696a335c9db6) - was part of a larger irrelevant commit, IIRC this is only a refactor and should not affect functionality

- [x] [4ecb54c](https://github.com/neovim/neovim/commit/4ecb54c16f5d8b0bceadf85cf90e14d032171379) - fixes an edge case where iirc fixes an **overly long** with `nowrap` enabled causes weird renderind
  
  - [x] (moved so that it directly follows 4ecb54c) [7d94ab9](https://github.com/neovim/neovim/commit/7d94ab93af7a3e6405f9bc2c5a8f5a9c98f0e9e0) - fixes multibyte characters from implementing the above commit
  - [x] [cc66bea](https://github.com/neovim/neovim/commit/cc66beacba68657bd1409cac581890dc1f918681) - The above commit, [4ecb54c](https://github.com/neovim/neovim/commit/4ecb54c16f5d8b0bceadf85cf90e14d032171379), sets `n_extra` to zero without selecting the next virtual text causing incorrect first character to be drawn

- [x] [d483e39](https://github.com/neovim/neovim/commit/d483e392339bd53e150b340eb41d1066fb6fbdb4) - only for virtual text in **diff mode**
- [x] (squashed into previous) [e844e92](https://github.com/neovim/neovim/commit/e844e92307acad859d541003dedb34c986427759) - only for virtual text in **diff mode**

- [x]  [6cb724d](https://github.com/neovim/neovim/commit/6cb724d304e2328b061368f1640809d91696eff4) - very specific edge case regarding tabs and **overly long** virtual text

- [x] (I fixed linting for every commit individually) [cc4f9e0](https://github.com/neovim/neovim/commit/cc4f9e0e234490eb38e5821483ffb61af99efafb) - linting (obv not required)

File: runtime/doc/news.txt [hunk 2]
```diff
@@ -52,6 +52,8 @@ iterators |luaref-in|.
 • |'smoothscroll'| option to scroll by screen line rather than by text line
 when |'wrap'| is set.
 
+• Added |"inline"| to the possible |'virt_text_pos'| values of |nvim_buf_set_extmark|
```
Comment (clason) [inline runtime/doc/news.txt:55]:
```suggestion
• Added inline virtual text support to |nvim_buf_set_extmark|.
```

File: runtime/doc/news.txt [hunk 3]
```diff
@@ -52,6 +52,8 @@ iterators |luaref-in|.
 • |'smoothscroll'| option to scroll by screen line rather than by text line
 when |'wrap'| is set.
 
+• Added inline virtual text support to |nvim_buf_set_extmark|.
```
Comment (clason) [inline runtime/doc/news.txt:55]:
```suggestion
• Added inline virtual text support to |nvim_buf_set_extmark()|.
```
(sorry, missed that the first time around)

**Comment by zeertzjq** [comment]
The first commit's message is missing two line breaks before the first URL pointing to the original Vim commit.

**Comment by tom-anders** [comment]
> The first commit's message is missing two line breaks before the first URL pointing to the original Vim commit.

Fixed, thanks

**Comment by theHamsta** [comment]
Just tested this feature with https://github.com/theHamsta/nvim-dap-virtual-text

image

What should be the meaning of start_col, start_row, end_col, end_row when this feature is used? Until now I set the range to the range of the node I want to annotate. I would understand that the function contract would say that `end_col`, `end_row` would be ignored in such case. But would it make sense to place the virtual text at `end_col`, `end_row` if specified so that the text placement would make sense for both `virt_text_pos = "eol"` and `virt_text_pos = "inline"` (Assuming that after the position/range is the default)?

image

At the moment, I have 
```
        vim.api.nvim_buf_set_extmark(buf, hl_namespace, node_range[inline and 3 or 1], node_range[inline and 4 or 2], {
          end_line = node_range[3],
          end_col = node_range[4],
          hl_mode = 'combine',
          virt_text = { virt_text },
          virt_text_pos = options.virt_text_pos,
          virt_text_win_col = options.virt_text_win_col and win_col,
        })
```
to support both `"eol"` and `"inline"` to annotate extmark ranges.

**Comment by clason** [comment]
@theHamsta (and whoever else is reading this): please avoid linking issues or PRs (or `@` mentioning people) in commit messages while developing; otherwise every force push will spam the linked issue/PR/person!

**Comment by bfredl** [comment]
>  @bfredl Anything we want to drop from the non-essential list?

I'd say lets keep all of them which you already have added to this branch.

I'll make a more through review tomorrow, but from a quick I think it looks reasonable to merge this quite soon, and then fix more issues on master (known bugs when using inline text is fine, as long as there is no known regressions with _not_ using it).

**Comment by bfredl** [comment]
@theHamsta for now only the "start" position is used which is something we could document. as a later follow up, we want to add a mode which allows you to _replace_ a range of buffer text with the virtual text in rendering (this would be the full "anti-conceal" feature)

File: src/nvim/buffer_defs.h [hunk 4]
```diff
@@ -808,6 +808,7 @@ struct file_buffer {
 
   MarkTree b_marktree[1];
   Map(uint32_t, uint32_t) b_extmark_ns[1];         // extmark namespaces
+  size_t b_virt_text;           // number of virtual texts
```
Comment (bfredl) [inline src/nvim/buffer_defs.h:811]: this should changed to `b_virt_text_inline` (and only count extmarks with inline virtual text)

**Comment by adelarsq** [comment]
> @theHamsta for now only the "start" position is used which is something we could document. as a later follow up, we want to add a mode which allows you to _replace_ a range of buffer text with the virtual text in rendering (this would be the full "anti-conceal" feature)

Just as a note. This would be cool to solve things like this:

image

File: src/nvim/buffer_defs.h (hunk 4 above)
Comment (tom-anders) [reply src/nvim/buffer_defs.h]: Done! 

For now I just added `&& decor->virt_text_pos == kVTInline` to the checks, but maybe it's worth introducing a helper function
 ```
bool decor_has_inline_virtual_text(Decoration *decor) { 
  return kv_size(decor->virt_text) && decor->virt_text_pos == kVTInline; 
}
```
for this?

**Comment by bfredl** [comment]
pushed a small fix to allow both left-gravity and right-gravity marks at the same position with virtual text.

**Comment by danilshvalov** [comment]
I'm not sure if this is a known bug, so I'm sorry if I'm giving unnecessary information. 

If you add virtual text in `inline` mode and then add virtual text in `overlay` mode at column `0`, then virtual text in `overlay` mode will be positioned incorrectly. If you put it at column `1`, then everything works as expected.

<details>
<summary>Example</summary>

1. Type `abc`
2. Execute
```lua
vim.api.nvim_buf_set_extmark(
  0,
  vim.api.nvim_create_namespace('test'),
  0,
  0,
  { virt_text_pos = 'inline', virt_text = { { '   ' } } }
)
```
3. Execute
```lua
vim.api.nvim_buf_set_extmark(
  0,
  vim.api.nvim_create_namespace('test'),
  0,
  0,
  { virt_text_pos = 'overlay', virt_text = { { 'A' } } }
)
```

### Result
Screenshot 2023-05-12 at 03 41 07

### Expected
Screenshot 2023-05-12 at 03 41 31

1. Type ` abc` (with space before `abc`)
2. Execute
```lua
vim.api.nvim_buf_set_extmark(
  0,
  vim.api.nvim_create_namespace('test'),
  0,
  0,
  { virt_text_pos = 'inline', virt_text = { { '   ' } } }
)
```
4. Execute (`col = 1` instead of `col = 0`)
```lua
vim.api.nvim_buf_set_extmark(
  0,
  vim.api.nvim_create_namespace('test'),
  0,
  1,
  { virt_text_pos = 'overlay', virt_text = { { 'A' } } }
)
```

### Result

Screenshot 2023-05-12 at 03 45 15

</details>

**Comment by danilshvalov** [comment]
Experimentally, I found out that the current code does not seem to add an win_col offset after using inline virtual text. I doubt this is a good solution, but it fixed this problem.

I also had to comment out one line not related to inline virtual text in order for everything to work. Maybe it will break something. I hope this solution will help you understand what the problem is and you will be able to solve this problem.

```patch
diff --git a/src/nvim/drawline.c b/src/nvim/drawline.c
index 2426728f7..e8e837609 100644
--- a/src/nvim/drawline.c
+++ b/src/nvim/drawline.c
@@ -1652,7 +1652,7 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
         wlv.draw_state = WL_LINE;
         if (has_decor && wlv.row == startrow + wlv.filler_lines) {
           // hide virt_text on text hidden by 'nowrap'
-          decor_redraw_col(wp, wlv.vcol, wlv.off, true, &decor_state);
+          // decor_redraw_col(wp, wlv.vcol, wlv.off, true, &decor_state);
         }
         win_line_continue(&wlv);  // use wlv.saved_ values
       }
@@ -1750,10 +1750,7 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
       }
 
       if (has_decor && v >= 0) {
-        bool selected = (area_active || (area_highlighting && noinvcur
-                                         && wlv.vcol == wp->w_virtcol));
-        extmark_attr = decor_redraw_col(wp, (colnr_T)v, wlv.off,
-                                        selected, &decor_state);
+        int col_off = 0;
 
         while (true) {
           // we could already be inside an existing virt_line with multiple chunks
@@ -1783,6 +1780,7 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
             VirtTextChunk vtc = kv_A(virt_inline, virt_inline_i);
             wlv.p_extra = vtc.text;
             wlv.n_extra = (int)strlen(wlv.p_extra);
+            col_off += wlv.n_extra;
             wlv.extra_for_extmark = true;
             wlv.c_extra = NUL;
             wlv.c_final = NUL;
@@ -1827,6 +1825,12 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
           }
           break;
         }
+
+        bool selected = (area_active || (area_highlighting && noinvcur
+                                         && wlv.vcol == wp->w_virtcol));
+        extmark_attr = decor_redraw_col(wp, (colnr_T)v, wlv.off + col_off,
+                                        selected, &decor_state);
+
       }
 
       if (wlv.n_extra == 0) {
```

**Comment by tom-anders** [comment]
> I'm not sure if this is a known bug, so I'm sorry if I'm giving unnecessary information.
> 
> If you add virtual text in `inline` mode and then add virtual text in `overlay` mode at column `0`, then virtual text in `overlay` mode will be positioned incorrectly. If you put it at column `1`, then everything works as expected.
> 
> UPD: Correction. If there is both a virtual text and an overlay in one column, it is displayed incorrectly. If different, then everything is fine.

@bfredl How do we want to go about this? Try to fix before merging? Or merge the PR and add a new issue for this bug?

File: src/nvim/drawline.c [hunk 5]
```diff
@@ -1723,7 +1749,87 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
         area_active = false;
       }
 
-      if (!wlv.n_extra) {
+      if (has_decor && v >= 0) {
```
Comment (lewis6991) [inline src/nvim/drawline.c:1752]: Can this whole block be factored into a function. `win_line()` is bad enough as it is.

**Comment by bfredl** [comment]
I plan to merge this later today after doing some cleanup myself. we can work on fixing more edge cases after that.

>  Can this whole block be factored into a function. win_line() is bad enough as it is.

I'll try.

**Comment by bfredl** [comment]
Included all changes into #20130

**Comment by nyngwang** [comment]
@bfredl You're awesome, thank you!

**Comment by SleepySwords** [comment]
If you didn't already know, you could use `right_gravity` option while inserting the inline virtual text. This makes it so that the text is inserted to the right instead of the left of the virtual text. I found this is particular nice for parameter type with some LSPs. For example, when calling Rust functions.

**Pull request closed without merging**